### PR TITLE
Rc/clickable icons request

### PR DESCRIPTION
**Pull request by Robert-Costello**
## Product Description

icon-click

## Technical Summary

This sets up the basic functionality of clickable icons, constructing the url based on the `url_template` in the entity list response and making an ajax request.

## Feature Flag

CASE_LIST_CLICKABLE_ICON

### Safety story

under feature flag 
tested locally
testing on staging and QA

### QA Plan

QA ticket: https://dimagi-dev.atlassian.net/browse/QA-5680

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by MartinRiese — APPROVED** [review]

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 1]
```diff
@@ -243,6 +245,40 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
             };
         },
 
+        iconClick: function (e) {
+            e.stopImmediatePropagation();
+            if (!e.target.className.includes('disabled')) {
```
Comment (Robert-Costello) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:250]: I'll also remove this check. Shouldn't be needed.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 2]
```diff
@@ -243,6 +245,39 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
+            const user = FormplayerFrontend.getChannel().request('currentUser');
+            const appId = currentUrlToObject.copyOf;
+            const domain = user.domain;
+            const caseId = this.model.get('id');
+            const urlTemplate = this.options.endpointActions[1]['urlTemplate'];
+            const actionUrl = origin + urlTemplate
+                .replace("{domain}", domain)
+                .replace("{appid}", appId)
+                .replace("{case_id}", caseId);
+            this.iconSmartLink(e, actionUrl);
```
Comment (Robert-Costello) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:261]: At the moment cases are not being updated with this call to ` this.iconSmartLink`, but if you copy `actionUrl` and paste it in the browser the form will be submitted and the case will be correctly updated. Looks like ajax request needs something else.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 3]
```diff
@@ -243,6 +245,39 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
+            const currentUrlToObject = formplayerUtils.currentUrlToObject();
+            const user = FormplayerFrontend.getChannel().request('currentUser');
+            const appId = currentUrlToObject.copyOf;
+            const domain = user.domain;
+            const caseId = this.model.get('id');
+            const urlTemplate = this.options.endpointActions[1]['urlTemplate'];
+            const actionUrl = origin + urlTemplate
+                .replace("{domain}", domain)
+                .replace("{appid}", appId)
+                .replace("{case_id}", caseId);
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:260]: This is probably OK for now but in future we may want to use something like https://www.npmjs.com/package/url-template

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 4]
```diff
@@ -243,6 +245,39 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
+            this.iconSmartLink(e, actionUrl);
+        },
+
+        iconSmartLink: function (e, url) {
+            e.target.className += " disabled";
+            self.processing = true;
+            return $.ajax({
+                type: "POST",
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:268]: does this not have to be a GET request?

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 5]
```diff
@@ -243,6 +245,40 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
 
+        iconClick: function (e) {
+            e.stopImmediatePropagation();
+            if (!e.target.className.includes('disabled')) {
+                const origin = window.location.origin;
+                const user = FormplayerFrontend.getChannel().request('currentUser');
+                let appId = FormplayerFrontend.getChannel().request('getCurrentAppId');
+                const domain = user.domain;
+                const caseId = this.model.get('id');
+                const urlTemplate = this.options.endpointActions[1]['urlTemplate'];
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: I'm not sure if you can do exactly this but here's how it works for sorting:
https://github.com/dimagi/commcare-hq/blob/759604624e42910892e80236f690b031ef4dfc3a/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js#L535

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 3 above)
Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:260]: Thanks, I didn't love this approach. I'll circle back and update it with a followup.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 4 above)
Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:268]: yes, it does.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: took care of this here: [ec6095e](https://github.com/dimagi/commcare-hq/pull/33597/commits/ec6095ea7bb72aa186e48deb4e295e9d4166b4a9)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 6]
```diff
@@ -243,6 +245,43 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
+            const domain = user.domain;
+            const caseId = this.model.get('id');
+            const fieldIndex = $(e.currentTarget).parent().index();
+            const urlTemplate = this.options.endpointActions[fieldIndex]['urlTemplate'];
+            const actionUrl = origin + urlTemplate
+                .replace("{domain}", domain)
+                .replace("{appid}", currentAppId)
+                .replace("{case_id}", caseId);
+            e.target.className += " disabled";
+            document.location = actionUrl;
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:265]: how come we are redirecting now?

Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:265]: this was an intermediate step. The background ajax request wasn't updating cases and was giving a false positive response. Working on the background request now, and I'll be removing the redirect next.

Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: fixed here https://github.com/dimagi/commcare-hq/pull/33597/commits/8781f14533f316d00fc9e04e704920d253b13423

**Review by MartinRiese — APPROVED** [review]

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 7]
```diff
@@ -243,6 +245,67 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
...
+                .replace("{domain}", domain)
+                .replace("{appid}", currentAppId)
+                .replace("{case_id}", caseId);
+            e.target.className += " disabled";
+            this.iconIframe(e, actionUrl);
+        },
+
+        iconIframe: function (e, url) {
+            const clickedIcon = e.target;
+            clickedIcon.classList.add("disabled");
```
[the remark this answers is not in the document]
Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:270]: The button gets enabled again [here](https://github.com/dimagi/commcare-hq/pull/33597/files#diff-59204083f0efa0e7d40586e28666900ebf9970de0550b5ffad1cd9f8fbc2c918R300:~:text=remove(%22-,disabled,-%22)%3B) after the form submission is complete. You can click the icon multiple times without leaving the page or refreshing, but this does force you to wait for the form the submit each time. Is that what you mean?

[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:270]: I did not realize it gets re-enabled. Disregard my comment then

**Pull request merged**